dynamyc_list/list.cpp

```diff
@@ -0,0 +1,417 @@
+	*list = NULL;
+}
+
+
+/* Inserts value at the end of the list. */
+void StringListAdd(char** list, String str)
+{
+	char** current_elem = list;
+	char** new_elem = (char**)malloc(2 * sizeof(char*));//create new elem
+	char* current_str = NULL;
+	current_str = (char*)malloc((strlen(str) + 1) * sizeof(char));
+	strcpy(current_str, str);
+	new_elem[0] = current_str;//add string to list
+	new_elem[1] = NULL;//set next elen as NULL
+
+	while (current_elem[1] != NULL)//search last elem
+	{
+		current_elem = (char**)current_elem[1];//go to next elem
+	}
+
+	if (current_elem[0] == NULL)
+	{		
+		current_elem[0] = new_elem[0];//add new elem to list
+	}
+	else
+	{
+		current_elem[1] = (char*)new_elem;//add new elem to list
+	}
+}
+
+/* Removes all occurrences of str in the list. */
+void StringListRemove(char** list, String str)
+{
+	char** current_elem = list;
+	int size = StringListSize(list);
+
+	int count_of_elem_with_str = 0;
+
+	if (strcmp(current_elem[0], str) == 0)
+	{
+		++count_of_elem_with_str;
+	}
+
+	char** previous_elem = current_elem;
+	current_elem = (char**)current_elem[1];
+
+	while (current_elem != NULL)//locking for and remove the elements with the same strings 'str'
+	{
+		if (strcmp(current_elem[0], str) == 0)//count elements with same strings 'str'
+		{
+			if (count_of_elem_with_str)
+			{
+				previous_elem[1] = current_elem[1];
+				free(current_elem);
+				current_elem = (char**)previous_elem[1];
+			}
+			++count_of_elem_with_str;
+		}
+		else
+		{
+			previous_elem = current_elem;
+			current_elem = (char**)current_elem[1];//go to next element
+		}
+	}
+}
+
+
+/* Returns the number of items in the list. */
+int StringListSize(char** list)
+{
+	char** current_elem = list;
+	int count = 0;//number of elements
+	while (current_elem != NULL)
+	{
+		++count;
+
+		current_elem = (char**)current_elem[1];//go to next element
+	}
+	return count;
+}
+
+/* Returns the index position of the first occurrence of str in the list. */
+int StringListIndexOf(char** list, char* str)
+{
+	char** current_elem = list;
+	int pos = 0;
+	if (str == NULL || current_elem == NULL)
+	{
+		return -1;
+	}
+	while (current_elem != NULL)
+	{
+		if (strcmp(current_elem[0], str)==0)
+		{
+			return pos;
+		}
+		else
+		{
+			++pos;
+		}
+
+		current_elem = (char**)current_elem[1];//go to next elemnt
+	}
+	return -1;
+}
+
+
+/* Removes duplicate entries from the list. */
+void StringListRemoveDuplicates(char** list)
+{
+	char** current_elem = list;
+	while (current_elem[1] != NULL)
+	{
+		StringListRemove(current_elem, current_elem[0]);
+	}
+}
+
+/* Replaces every occurrence of the before, in each of the string lists's strings, with after. */
+void StringListReplaceInStrings(char** list, char* before, char* after)
+{
+	char** current_elem = list;
+	if (after == NULL || current_elem == NULL)
+	{
+		return;
+	}
+	while (current_elem != NULL)
+	{
+		if (strcmp(current_elem[0], before) == 0)
+		{			
+			char* str = (char*)malloc((strlen(after) + 1) * sizeof(char));
```

Comment:
Line 162 Security issue
Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected)

Reply:
> Line 162 Security issue
> Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected)

strlen() return string's size without \0, but for free() \0 naad at the string's end for correctly deletet it
Because it i add 1 to strlen() for \0